dev/src/dev/rewrite_defendpoint_1.clj

```diff
@@ -0,0 +1,422 @@
+                            [(n/map-node
+                              ;; apparently the old behavior for booleans coerced `nil`, to `false`, so let's replicate
+                              ;; that.
+                              (if (= (n/sexpr schema) [:maybe 'ms/BooleanValue])
+                                [(n/keyword-node :default) (n/whitespace-node " ") (n/token-node 'false)]
+                                [(n/keyword-node :optional) (n/whitespace-node " ") (n/token-node 'true)]))
+                             (n/whitespace-node " ")])
+                          [schema]))])))
+           schema-map))))
+
+(defn- select-keys*
+  "Normal Clojure select-keys does not preserve the original map type e.g. ordered-map."
+  [m ks]
+  (when m
+    (into (empty m)
+          (filter (let [key-set (set ks)]
+                    (fn [[k v]]
+                      (when (key-set k)
+                        [k v]))))
+          m)))
+
+(mu/defn- keys-map :- ::node
+  [args :- [:sequential symbol?]]
+  (n/map-node [(n/keyword-node :keys)
+               (n/whitespace-node " ")
+               (n/vector-node (into []
+                                    (comp (map n/token-node)
+                                          (interpose (n/whitespace-node " ")))
+                                    args))]))
+
+(mu/defn- unused-binding? [node :- ::node]
+  (and (= (n/tag node) :token)
+       (str/starts-with? (n/sexpr node) "_")))
+
+(mu/defn- schema-specifier :- [:maybe [:sequential ::node]]
+  [schema :- [:maybe ::node]]
+  (when schema
+    [(n/whitespace-node " ")
+     (n/keyword-node :-)
+     (n/whitespace-node " ")
+     schema]))
+
+;;;
+;;; route params
+;;;
+
+(declare query-args-binding
+         body-binding)
+
+(mu/defn- route-args-symbols :- [:sequential symbol?]
+  [{:keys [route], :as _parsed}]
+  (let [route (if (= (n/tag route) :vector)
+                (-> route z/of-node z/down z/node)
+                route)
+        route (n/sexpr route)]
+    (assert (string? route)
+            (format "Expected route string, got %s" (pr-str route)))
+    (map (comp symbol second) (re-seq #":([^:/]+)" route))))
+
+(mu/defn- route-args-schema :- [:maybe ::node]
+  [{:keys [schema-map], :as parsed} :- ::parsed]
+  (some-> schema-map
+          (select-keys* (route-args-symbols parsed))
+          ;; route args cannot be optional so fix incorrect schemas if we see them.
+          (update-vals (fn [schema]
+                         (if (and (= (n/tag schema) :vector)
+                                  (= (first (n/sexpr schema)) :maybe))
+                           (z/node (-> (z/of-node schema) z/next z/next))
+                           schema)))
+          schema-map->malli))
+
+(mu/defn- route-args-binding :- ::node
+  [parsed]
+  (if-let [args (not-empty (route-args-symbols parsed))]
+    (keys-map args)
+    (n/token-node '_route-params)))
+
+(mu/defn- new-route-param-arg-nodes :- [:maybe [:sequential ::node]]
```

Comment:
with `defendpoint` 2 the route/query/body params bindings are all optional, if they're unused you don't need to bother putting `_whatever` for them so this code leaves them out for stuff like `GET /api/whatever/` that doesn't accept any parameters